### DISCUSSION TITLE
Handling OPTIONS request with mongoose

### DISCUSSION THREAD

**Discussion by NRollo** [cesanta/mongoose#1443] [Ask questions, get answers] [answered]
I was wondering, what is the recommended procedure for handling OPTIONS requests (preflight) with the mongoose webserver?
At the moment I am just rejecting this type of requests with a 501 "Not Implemented".

I am thinking of a response like:
`mg_http_reply(nc, 204, "Allow: OPTIONS, GET, POST\r\n", "");`

Thanks!

**Comment by cpq** [comment, marked as answer]
Yeah that sounds good. Something like this:

```c
static void fn(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
  if (ev == MG_EV_HTTP_MSG) {
    struct mg_http_message *hm = (struct mg_http_message *) ev_data;
    if (mg_strcmp(hm->method, mg_str("OPTIONS")) == 0) {
      mg_http_reply(c, 204, "Allow: OPTIONS, GET, POST\r\n", "");
    } else if (mg_http_match_uri(hm, "/api/foo")) {
      mg_http_reply(c, 200, "", "{\"result\": %d}\n", 123);  // Serve REST
    } else {
      struct mg_http_serve_opts opts = {.root_dir = "."};
      mg_http_serve_dir(c, ev_data, &opts);
    }
  }
}
```

**Comment by NRollo** [comment]
@cpq Thanks for the reply - could you elaborate a bit on the last part of the code:
```
} else if (mg_http_match_uri(hm, "/api/foo")) {
      mg_http_reply(c, 200, "", "{\"result\": %d}\n", 123);  // Serve REST
    } else {
      struct mg_http_serve_opts opts = {.root_dir = "."};
      mg_http_serve_dir(c, ev_data, &opts);
    }
```
Thanks!

**Reply by cpq** [reply]
The event handler implements the following functionality:
1. Respond to the OPTIONS request, regardless of the requested URI
2. Respond to `/api/foo` requests with a JSON string
3. Serve a directory for any other requested URI
